FlowDroid: Clears für Setter nachtragen

iricarte opened the pull request:
- Clears für mehrere Setters.
- Wichtige Annahme: Nur Setters mit Signatur 'void setX' haben Clear-Einträge bekommen. Setters, die bereits Clears hatten, wurden ignoriert.

iricarte committed:
FlowDroid: Clears für Setter nachtragen

- Clears für Setters.
- Wichtige Annahme: Jeder Setter mit Signatur 'void set...' hat einen
Clear-Eintrag bekommen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/org.xml.sax.XMLReader.xml:
```diff
@@ -102,7 +102,15 @@
 		<method id="void setFeature(java.lang.String,boolean)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Field"
```
Da ist eine Map-artige Struktur. Wenn ein Wert geschrieben wird, kannst du nicht alle anderen Werte ebenfalls wegwerfen. Schau mal in die ``java.util.Map``, wie das dort gemacht wird. Vermutlich fehlt hier einfach ein Constraint.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.util.concurrent.atomic.AtomicReferenceArray.xml:
```diff
@@ -58,7 +58,12 @@
 		</method>
 		<method id="void set(int,java.lang.Object)">
-			<flows>
+			<clears>
```
Du setzt ein Element in einem Array. Deswegen kannst du nicht alle Taints auf dem Array wegwerfen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.util.ListIterator.xml:
```diff
@@ -36,7 +36,11 @@
 		<method id="void set(java.lang.Object)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Field" AccessPath="[java.util.Iterator: java.lang.Object[] innerArray]"
```
Ein Element in einem Iterator zu ändern darf nicht die Taints von allen anderen Elementen in der zugrundeliegenden Collection entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/android.util.SparseArray.xml:
```diff
@@ -112,7 +112,12 @@
         <method id="void setValueAt(int,java.lang.Object)">
-            <flows>
+            <clears>
+            	<clear sourceSinkType="Field" BaseType="android.util.SparseArray"
```
Hier fehlt ein Constraint. Du kannst nicht alle Taints von einem Array löschen, wenn einzelnes Element überschrieben wird.

StevenArzt commented on soot-infoflow-summaries/summariesManual/android.util.JsonWriter.xml:
```diff
@@ -64,7 +64,11 @@
 		</method>
 		<method id="void setIndent(java.lang.String)">
-			<flows>
+			<clears>
```
Prüfe noch einmal, ob diese Methode wirklich die ganze Daten im Writer überschreibt. Ich habe da Zweifel.

iricarte committed:
FlowDroid: Clears für Setter nachtragen

- JsonWriter: Clear entfernt, da es nicht die ganze Datei überschreibt.
- XMLRader: Constraint hinzugefügt
- SparseArray: Constraint hinzugefügt
- AtomicReferenceArray: Constraint hinzugefügt
- ListIterator: Clear entfernt

Nach zusätzlicher Analyse sind mir noch folgende Anpassungen
aufgefallen:
- Address: Constraint für setAddressLine

StevenArzt commented on soot-infoflow-summaries/summariesManual/android.graphics.Bitmap.xml:
```diff
@@ -569,7 +599,18 @@
         <method id="void setPixel(int,int,int)">
-            <flows>
+            <clears>
+            	<clear sourceSinkType="Field" BaseType="android.graphics.Bitmap"
```
Das ist drei Mal drin, was keinen Sinn ergibt. Zudem ist es auch nicht sinnvoll, alle Taints auf der Bitmap zu clearen, wenn eine Koordinate überschrieben wird.

iricarte replied:
Dabei dann Constraints auf die Koordinaten hinzufügen, oder?

StevenArzt replied:
Wir unterstützen bei den Constraints keine Bedingungen über mehrere Parameter hinweg. Der Constraint müsste für das Paar ´´(x,y)`` zutreffen und das kann unsere Implementierung nicht. Also müssen wir hier über-approximieren und auf den Clear verzichten.

StevenArzt commented on soot-infoflow-summaries/summariesManual/android.graphics.Bitmap.xml:
```diff
@@ -591,7 +632,12 @@
         <method id="void setPixels(int[],int,int,int,int,int,int)">
-            <flows>
+            <clears>
+            	<clear sourceSinkType="Field" BaseType="android.graphics.Bitmap"
```
Die ganze Bitmap zu clearen wenn ein Teil davon überschrieben wird ist nicht sinnvoll

StevenArzt commented on soot-infoflow-summaries/summariesManual/android.util.SparseArray.xml:
```diff
@@ -112,7 +112,17 @@
+                        AccessPath="[android.util.SparseArray: byte[] data]"
+                        AccessPathTypes="[byte[]]" 
+                        constrained="true" 
+                        isAlias="true" />
```
Können Clears Aliase haben? Da ist mir die Semantik gerade nicht klar.

iricarte replied:
Ich glaube schon. Analog zu put(java.lang.Object,java.lang.Object) in Map.

StevenArzt commented on soot-infoflow-summaries/summariesManual/jdk.internal.misc.Unsafe.xml:
```diff
@@ -311,7 +311,10 @@
 		
 		<method id="void setMemory(java.lang.Object,long,long,byte)">
-			<flows>
+			<clears>
```
Da wir nicht wissen, welcher Teil vom Speicher überschrieben wird, würde hier nicht clearen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/jdk.internal.misc.Unsafe.xml:
```diff
@@ -320,7 +323,10 @@
 		
 		<method id="void setMemory(long,long,byte)">
-			<flows>
+			<clears>
```
Da wir nicht wissen, welcher Teil vom Speicher überschrieben wird, würde hier nicht clearen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/sun.misc.Unsafe.xml:
```diff
@@ -294,7 +297,10 @@
 		
 		<method id="void setMemory(long,long,byte)">
-			<flows>
+			<clears>
```
Da wir nicht wissen, welcher Teil vom Speicher überschrieben wird, würde hier nicht clearen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/sun.misc.Unsafe.xml:
```diff
@@ -285,7 +285,10 @@
 		
 		<method id="void setMemory(java.lang.Object,long,long,byte)">
-		<flows>
+		<clears>
```
Da wir nicht wissen, welcher Teil vom Speicher überschrieben wird, würde hier nicht clearen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.StringBuilder.xml:
```diff
@@ -399,7 +399,11 @@
 		</method>
 		<method id="void setCharAt(int,char)">
-			<flows>
+			<clears>
```
Hier wird nur ein einzelner Character ersetzt. Da können wir nicht den Taint von allen Characters löschen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		</method>
 		<method id="void set(java.lang.Object,int,java.lang.Object)">
-			<flows>
+			<clears>
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		</method>
 		<method id="void setBoolean(java.lang.Object,int,boolean)">
-			<flows>
+			<clears>
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setByte(java.lang.Object,int,byte)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setChar(java.lang.Object,int,char)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setShort(java.lang.Object,int,short)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setInt(java.lang.Object,int,int)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setLong(java.lang.Object,int,long)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setFloat(java.lang.Object,int,float)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.reflect.Array.xml:
```diff
@@ -74,71 +74,98 @@
 		<method id="void setDouble(java.lang.Object,int,double)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Parameter" ParameterIndex="0" />
```
Es wird nur ein Element des Arrays überschrieben. Deshalb können wir nicht alle Taints vom gesamtem Array entfernen.

StevenArzt commented on soot-infoflow-summaries/summariesManual/java.lang.StringBuffer.xml:
```diff
@@ -429,7 +429,11 @@
 		<method id="void setCharAt(int,char)">
-			<flows>
+			<clears>
+				<clear sourceSinkType="Field" 
```
Es wird nur ein einzelnes Zeichen ersetzt. Da können wir nicht alle Taints auf dem gesamten StringBuffer entfernen.

iricarte committed:
FlowDroid: Clears für Setter nachtragen

- Constraints für index-basierte Setters
- Löschen von nicht unterstützten Clears.
- Fix SparseArray AccessPath und AccessPathTypes
